src/main/java/uk/gov/companieshouse/model/validator/CompanyNumberValidator.java

```diff
@@ -0,0 +1,65 @@
+package uk.gov.companieshouse.model.validator;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import org.springframework.stereotype.Component;
+
+import uk.gov.companieshouse.config.CompanyNumberConfiguration;
+
+@Component
+public class CompanyNumberValidator implements ConstraintValidator<ValidCompanyNumber, String> {
+
+    private final Pattern prefixListRegex = Pattern.compile("(?i)^([A-Z][A-Z]?)\\b(,[A-Z][A-Z]?)*$");
+
+    private final Pattern companyNumberRegex;
+
+    public CompanyNumberValidator(CompanyNumberConfiguration prefixConfig) {
+
+        String prefixString = prefixConfig.getPrefixes();
+
+        if (!this.prefixListRegex.matcher(prefixString).matches()) {
+            throw new IllegalArgumentException(
+                    "Prefix list formatting error. Make sure list is comma separated e.g. NI,SI,R. Current: "
+                            + prefixString);
+        }
+
+        List<String> prefixesArray = Arrays.asList(prefixString.split(","));
+
+        String singleCharacterPrefixRegex = "("
+                .concat(createPrefixSubstring(prefixesArray, 1).concat(")").concat("[0-9]{1,7}"));
+
+        String doubleCharacterPrefixRegex = "("
+                .concat(createPrefixSubstring(prefixesArray, 2).concat(")").concat("[0-9]{1,6}"));
+
+        String onlyNumbersRegex = "[0-9]{1,8}";
+
+        this.companyNumberRegex = Pattern.compile("(?i)^("
+                + String.join("|", List.of(singleCharacterPrefixRegex, doubleCharacterPrefixRegex, onlyNumbersRegex))
+                + ")$");
+
+    }
+
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        return this.companyNumberRegex.matcher(value).matches();
```

Comment:
I would advise breaking this up so it's more readable, example below (variable and method names could be improved :P):

```
    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {

        boolean prefixIsValid = arePrefixesValid();
        boolean companyNumberIsValid = isCompanyNumberValid(value);

        return prefixIsValid && companyNumberIsValid;
    }
```